web/app/components/input/input.tsx

```diff
@@ -148,10 +172,163 @@ export class Input extends Component<Props, State> {
       });
   }
 
-  render(props: RenderableProps<Props>, { isDisabled, isErrorShown, errorMessage, preview, maxLength, text }: State) {
+  /** appends error to input's error block */
+  appendError(...errors: string[]) {
+    if (!this.state.errorMessage) {
+      this.setState({
+        errorMessage: errors.join('\n'),
+        isErrorShown: true,
+      });
+      return;
+    }
+    this.setState({
+      errorMessage: this.state.errorMessage + '\n' + errors.join('\n'),
+      isErrorShown: true,
+    });
+  }
+
+  onDragOver(e: DragEvent) {
+    if (StaticStore.config.max_image_size === 0) return;
+    if (!this.textAreaRef) return;
+    if (!e.dataTransfer) return;
+    const items = Array.from(e.dataTransfer.items);
+    if (Array.from(items).filter(i => i.kind === 'file' && ImageMimeRegex.test(i.type)).length === 0) return;
+    e.preventDefault();
+    e.dataTransfer.dropEffect = 'copy';
+  }
+
+  async onDrop(e: DragEvent) {
+    if (StaticStore.config.max_image_size === 0) return;
+    if (!this.textAreaRef) return;
+    if (!e.dataTransfer) return;
+
+    const data = Array.from(e.dataTransfer.files).filter(f => ImageMimeRegex.test(f.type));
+    if (data.length === 0) return;
+
+    e.preventDefault();
+
+    /** Human readable image size limit, i.e 5MB */
+    const maxImageSizeString = (StaticStore.config.max_image_size / 1024 / 1024).toFixed(2) + 'MB';
+    /** upload delay to avoid server rate limiter */
+    const uploadDelay = 5000;
+
+    const isSelectionSupported = this.textAreaRef.isSelectionSupported();
+
+    this.setState({
+      errorLock: true,
+      errorMessage: null,
+      isErrorShown: false,
+      isDisabled: true,
+      buttonText: 'Uploading...',
+    });
+
+    // fallback for ie < 9
+    if (!isSelectionSupported) {
+      for (let i = 0; i < data.length; i++) {
+        const file = data[i];
+        const isFirst = i === 0;
+        const placeholderStart = this.state.text.length === 0 ? '' : '\n';
+
+        if (file.size > StaticStore.config.max_image_size) {
+          this.appendError(`${file.name} exceeds size limit of ${maxImageSizeString}`);
+          continue;
+        }
+
+        !isFirst && (await sleep(uploadDelay));
```

Comment:
`sleep` for me it is `wait`

Reply:
and also `delay`, what purpose of renaming?

Reply:
I like `delay` too.
`sleep` has connotation in others language. Is is about suspends the execution of the current thread until the timeout interval elapsed.

Reply:
I think logic about sequence upload file could be moved to UploadService (or something like this). It should not be in presentation level.
Let's imagine, if we add file input for upload, we need copy the same login from DND file.

Reply:
If you move it to service you could add unit tests for conditions.

Reply:
That can be made not as HOC, but additional function which takes `File[]`, and processes them. Seems reasonable as after #305 this button should be added as well.

As for service, right now it tightly coulpled with ui, it constantly changes textarea value, so I don't think it's reasonable. You can add unit tests by mocking `Props.uploadImage`

Reply:
> As for service, right now it tightly coulpled with ui
You are right. And it problem. We can create service which will notify about progress of upload files and in component we subscribe on that events. The the simple way it trigger redux action and add midleware.

Reply:
As for `sleep` vs `delay`, it's so minor, that it pointless to discuss

Reply:
> You are right. And it is problem. We can create service which will notify about progress of upload files and in component we subscribe on that events. The the simple way it trigger redux action and add midleware.

If there will be demand, sure, right now uploading presented only in Input component.

> The the simple way it trigger redux action and add midleware.
Nah, it's not simple, what is the point of redux and middleware here?

Reply:
If you want decouple data and presentation layer. You need implement `Observer` for this case. But as we use redux for state management now we can use it for upload files.
how it could be look:
```js
//action Creator
function(files) {
  dispatch({type: 'UPLOAD_FILES_REUEST', PAYLOAD: {}})
  for(const file of files) {
      dispatch({type: 'UPLOAD_FILE_REUEST', payload: {fileName: 'fileName'}})
      fetcher
     .uploadFile(file)
     .then((res) => {
           dispatch({type: 'UPLOAD_FILE_REUEST_COMPLETED', payload: {fileName: 'fileName'}})
      })
     .then((res) => {
           dispatch({type: 'UPLOAD_FILE_REUEST_FAILED', payload: {fileName: 'fileName'}})
      })
     .then(() => {
          return delay(1500);
     });
  }
 dispatch({type: 'UPLOAD_FILES_REUEST_COMPLETED', PAYLOAD: {}})
}

```
```js
// in store we add state
{
...
files:{...}
}
```
And connect this state to input.
Yes middleware don't need here.
And this code will be easy reuse for upload from file input

Reply:
IMHO. It is not about requirements. It is more about readability and maintainability.

Reply:
what purpose `store.files` has?

Action you provided does different logic. It delays after every upload, but I need to delay only if there are several items, item is not first and on start of upload. And I don't even understand how to use it, should I use `store.files` as a queue? Hardly doubt that it increases readability and maintainability.

as for file input, it can be done easily as
```
class Input extends ... {
  ...
  onFileInputChange(e){
    const files = filterImages(Array.from(e.target.files))
    if (files.length === 0) return;
    this.uploadImages(files)
  }
  ...
}